src/python/pants/engine/legacy/graph.py

```diff
@@ -393,7 +393,13 @@ def __hash__(self):
     def __eq__(self, other):
         if not isinstance(other, HydratedTarget):
             return NotImplemented
-        return self._address == other._address
+        # NB: It's an anti-pattern for `__hash__` to be implemented differently than `__eq__`. We
```

Comment:
This could lead to performance problems of another kind though.  I.e., constant-time hash lookups turn into O(n) hash bucket scans, because many versions of the same target end up hashed together into the same bucket.  And that would be subtle and creep up slowly, and be hard to debug. 

So I feel like this is not ideal. Could the performance issue instead be mitigated by memoizing the hash? Or is even the one-time computation too expensive?

Reply:
FTR, memoizing resulted in worse performance. I found that the current setup had as good performance as before this PR.